### ISSUE TITLE
Home Assistant 2022.2 not starting

### ISSUE THREAD

**Issue by piereligio** [tprelog/iocage-homeassistant#52]
I don't really know where to ask, I'm not sure if this issue is because of the integration or of Home Assistant Core itself.
Basically what happens is that if I try to upgrade homeassistant service from the shell of this integration, what happens is that homeassistant service won't start. It prints "Home assistant only supports Linux, OSX and Windows using WSL", then it says it is launched with the local URL where I should be able to open the GUI, but it will never be accessible. If I check the status of the service giving input 2, it will say that homeassistant isn't running, and if I restart, it tells me something like "home assistant not running check homeassistant_daemon.pid" .

Basically the question if it is only matter of time before 2022.2 will get supported, or if it's my installation that got bugged up somehow, or if I have something to edit in the configuration files.

Thanks in advance.

**Comment by tprelog** [comment]
> . It prints "Home assistant only supports Linux, OSX and Windows using WSL",

Oh :hankey:  I'm in the process of spinning up a TrueNAS CORE instance to have a look

**Comment by piereligio** [comment]
Thank you very much 😁

**Comment by adorobis** [comment]
Same here. Had to roll back to snapshot from before upgrade...

**Comment by adorobis** [comment]
On the release party on youtube channel they've just confirmed the fix - there is a supported os check included by default which can be overridden by a command line parameter. @tprelog is about to implement a fix :) This is fast support :)

**Comment by adorobis** [comment]
> snapshots were a godsend

Yeah, I have it as a habit since a few months to always take a snapshot just before any upgrades...

**Comment by piereligio** [comment]
> > snapshots were a godsend
> 
> Yeah, I have it as a habit since a few months to always take a snapshot just before any upgrades...

I have periodic snapshots everyday for the whole pool, so I can always use those even in the worst case possible (which is me forgetting to do important stuff lol).


> On the release party on youtube channel they've just confirmed the fix - there is a supported os check included by default which can be overridden by a command line parameter. @tprelog is about to implement a fix :) This is fast support :)

Being it a known command line parameter, is it something I can do even before a new release of this integration? I mean, not that I'm so in a hurry, but it would be great. I can't wait to set up local Google Assistant feature, you know ahah

**Comment by tprelog** [comment]
Ok here's the fix

```bash
iocage console $jail_name
#Press 0 to exit menu

ee /usr/local/etc/rc.d/homeassistant
```

Edit line `116` of the script to add `--ignore-os-check` as shown below

```bash
HA_ARGS="--ignore-os-check --config ${homeassistant_config_dir}"
```

Press `esc` then `enter` twice to save and exit

Finally start  Home Assistant

```bash
service homeassistant start
```

**Comment by piereligio** [comment]
> Ok here's the fix
> 
> ```shell
> iocage console $jail_name
> #Press 0 to exit menu
> 
> ee /usr/local/etc/rc.d/homeassistant
> ```
> 
> Edit line `116` of the script to add `--ignore-os-check` as shown below
> 
> ```shell
> HA_ARGS="--ignore-os-check --config ${homeassistant_config_dir}"
> ```
> 
> Press `esc` then `enter` twice to save and exit
> 
> Finally start Home Assistant
> 
> ```shell
> service homeassistant start
> ```

It's working, thank you so much, amazing support! 😁

**Comment by adorobis** [comment]
Many thanks for super fast fix! 
> Edit line `116`

In my case it was line 146 and I don't recall to edit the file before.
Btw, when checking the config from command line menu it also shows the same message and I think it is not performing the config check.
![image](https://user-images.githubusercontent.com/19752098/152245812-7f4b1f42-f07a-4cb4-ac66-6e3203b5e77d.png)

**Comment by piereligio** [comment]
> Many thanks for super fast fix!
> 
> > Edit line `116`
> 
> In my case it was line 146 and I don't recall to edit the file before. Btw, when checking the config from command line menu it also shows the same message and I think it is not performing the config check. ![image](https://user-images.githubusercontent.com/19752098/152245812-7f4b1f42-f07a-4cb4-ac66-6e3203b5e77d.png)

The line will still appear, but if you check status, it won't stop running now. If you wait long enough, GUI will be available at your usual local address.

**Comment by adorobis** [comment]
Yes, HA starts although it took a very long time. Also need to update the zwavejs2mqtt plugin as it requires latest version now (and this one also takes quite a long time). Let's see if anything else is broken...

**Comment by tprelog** [comment]
> In my case it was line 146 and I don't recall to edit the file before.

@adorobis you must be on an older version of the plugin -- recent version should show the plugin version on the main menu

![image](https://user-images.githubusercontent.com/35702532/152248857-23cfe5a6-3926-44db-af49-50bfdad159ca.png)

**Comment by tprelog** [comment]
Plugin is updated now. Running a Plugin UPDATE will apply required changes.

**Comment by adorobis** [comment]
> you must be on an older version of the plugin

Indeed, I don't remember when doing the last update, for sure before November last year. I think I've only tried to update the plugin once in the past and it failed so was rather not doing it later. Is the plugin snapshot also allowing to roll back to before the plugin Update?

**Comment by tprelog** [comment]
> I don't remember when doing the last update,

I think you can still see your version in the POST INSTALL NOTES
![image](https://user-images.githubusercontent.com/35702532/152251201-ec26060e-4014-4f20-957f-a879ce3bc977.png)

Getting a bit off topic, this is kinda an area I'm torn between just leaving well enough alone or some further development effort. For example I've debated in my head if we should have a way to run an update from the menu, that will only update files contained in the [plugin overlay](https://github.com/tprelog/iocage-homeassistant/tree/master/overlay). Aside from changes to the plugin manifest, which are less frequent, that would essentially allow you to keep the "plugin"  updated without running a Plugin UPDATE.

**Comment by piereligio** [comment]
Follow up on the original issue: basically it seems to happen still, when I restart home assistant server from home assistant GUI. It's not a big issue, since it doesn't happen when restarting the whole jail, but I think it made sense to report it anyways.

**Comment by adorobis** [comment]
Yeah, pretty old here: `version: v_0.4.0` and the release is `12.1-RELEASE-p13`
I'm hesitating to update as some of my plugins fail to update.
Would plugin Update without selecting to update the jail update only the plugin overlay files?

**Comment by adorobis** [comment]
> it doesn't happen when restarting the whole jail

Yeah, I'm always restarting entire jail as some integrations don't reload properly when just restarting the service or from GUI (e.g. yamaha integration seems to keep some tcp connections open)

**Comment by tprelog** [comment]
> it seems to happen still, when I restart home assistant server from home assistant GUI.

Unfortunately, I don't think I can affect how the GUI works. It is possible, (already a good case for the new button entity helper) to trigger the `service homeassistant restart` from within HA.

You could probably restart the entire jail as well but that will be a little more involved, getting into using the TrueNAS API

**Comment by tprelog** [comment]
> Would plugin Update without selecting to update the jail update only the plugin overlay files?

No but you can it manually for now... If you'd like, let's open a separate discussion about it

**Comment by tprelog** [comment]
@piereligio - think we have a workaround for the restart issue. I've already updated the plugin

https://github.com/tprelog/iocage-homeassistant/releases/tag/v7-2022.2.3

If you'd rather update your rc script manually instead, [just apply these changes](https://github.com/tprelog/iocage-homeassistant/commit/10910f3c7ec97f48d95bc68d102505bdac5a60d5) like we did above

**Comment by piereligio** [comment]
> @piereligio - think we have a workaround for the restart issue. I've already updated the plugin
> 
> https://github.com/tprelog/iocage-homeassistant/releases/tag/v7-2022.2.3
> 
> If you'd rather update your rc script manually instead, [just apply these changes](https://github.com/tprelog/iocage-homeassistant/commit/10910f3c7ec97f48d95bc68d102505bdac5a60d5) like we did above

Thank you very much 😁

**Comment by b-neufeld** [comment]
Ran updates today and can't get back into the GUI after quite a bit of waiting... 

Does this message mean something is still migrating in the background that I have to wait to complete? 
![image](https://user-images.githubusercontent.com/18536548/152839827-c65e2438-1ad1-4315-8c0a-576b8722f721.png)

Thanks for the discussion & notes so far.

**Comment by adorobis** [comment]
It took longer than usual to start after upgrading to 2022.2.0 In the mean time check if the service is running (option 2 on that menu screen) and if there are any messages in the log file (`/home/homeassistant/homeassistant/home-assistant.log`)

**Comment by b-neufeld** [comment]
Thanks for the reply! How much longer - like half an hour or more? If so, I will let it sit all afternoon. 

Option 2: 
> homeassistant is running as pid 48438.
> http://172.16.1.108:8123

Nothing exciting in the logfile. It appears to have started up successfully at least once after upgrading because it was griping about a Volumio device loading slowly. Every subsequent restart of the service or reboot of the jail, nothing happens. 

The log file was >300 MB so I renamed it and created a new logfile "home-assistant.log" and ensured it had the right permissions and owner. When I restart the home assistant service, the logfile stays at filesize zero.  

As a sanity-check I searched the jail for other instances of "home-assistant.log" (found none) and double-checked the mount point mapping (appears unchanged). 

I tried running the upgrader again from the console, all requirements are already satisfied, and it signs off with these lines again:
> Testing configuration at /home/homeassistant/homeassistant
> INFO:homeassistant.helpers.storage:Migrating core.device_registry storage from 1.1 to 1.2
> INFO:homeassistant.helpers.storage:Migrating core.entity_registry storage from 1.1 to 1.4

... and logfile remains empty. 

Running service -e in the console, I also see "/usr/local/etc/rc.d/homeassistant" running. 

Any thoughts are appreciated. The empty logfile is strange. And sorry if this is not the right place to troubleshoot (please advise if not), but it seemed relevant.

**Comment by tprelog** [comment]
Hi @b-neufeld sorry for the trouble, but glad to hear you got it sorted... and thank you @adorobis for jumping in

Just a quick follow up here

> Everything has been booting up for 20 minutes

There is at least one report of someone waiting around 45 minutes for Home Assistant to come back after an update... But that was after a Plugin UPDATE which reinstalled Home Assistant during the process. (That's when I decided I would not reinstall home assistant automagically anymore on future updates, even when I know it needs to be.)

>The log file was >300 MB so I renamed it and created a new logfile "home-assistant.log" and ensured it had the right permissions and owner. When I restart the home assistant service, the logfile stays at filesize zero.

That could be normal - The default log level in HA is set to warning so we typically don't see anything until it breaks. I was just sharing with @adorobis [in this discussion](https://github.com/tprelog/iocage-homeassistant/discussions/54#discussioncomment-2129838), some ideas to change the log levels `info` and get more feedback during an update

**Comment by UraniumDonut** [comment]
https://github.com/home-assistant/core/pull/64352 has to be reverted or changed to support FreeBSD. Home-Assistant works and is actively supported for FreeBSD, yet the check does not account for it. Releasing an update into the main branch into an established program, which breaks thousands of installs in not acceptable @emontnemery

**Comment by adorobis** [comment]
> Home-Assistant works and is actively supported for FreeBSD

I'm afraid it isn't. Neither officially nor actively supported. Point me to the place where it states otherwise.

**Comment by tprelog** [comment]
Sorry @UraniumDonut  but FreeBSD has never been supported by Home Assistant. We are lucky that for the most part it just works.

Please see this Home Assistant ADR for details
https://github.com/home-assistant/architecture/blob/6da4482d171f2ef04de9320d313526653b5818b4/adr/0016-home-assistant-core.md#0016-installation-method-home-assistant-core

**Comment by aliekiddo** [comment]
I'm having the same issue after upgrading to 2022.2.3 - I previously had 2021.3.4 which stopped successfully restarting with the WebUI in HA, only when restarting the jail would it come back up.

I thought updating the plugin / upgrading with the shell would help (I've tried both) but I've not been able to get the UI working for several hours.
I've tried restarting from shell; restarting the jail; stopping and starting; rebooting TrueNAS. It's been at least 1:30 after I've rebooted and it's still down.

**Comment by tprelog** [comment]
Hi @aliekiddo - What you are facing is likely more than this issue alone, which is caused by a new OS check implemented by Home Assistant.

> after upgrading to 2022.2.3 - I previously had 2021.3.4

That's a really big jump for Home Assistant and this plugin. I imagine there are a few breaking changes you'll need to address with Home Assistant, including the required (yearly) python upgrade.

If you are using an external dataset for your configuration, I would suggest just doing a clean install of the plugin and remount your configuration there.

Otherwise you'll first need to run a Plugin UPDATE from the TrueNAS UI. This should ensure you should have the new python version installed as well as other file that have since been updated in this plugin, including the fix for the OS check. 

Next you'll need to manually switch python versions and rebuild the virtualenv for Home Assistant

```bash
iocage console $JAIL_NAME
# press 0 to exit menu
sysrc homeassistant_python=/usr/local/bin/python3.9
service homeassistant reinstall homeassistant
```

Also it will take some time for Home Assistant to reinstall itself. It will still be downloading and compiling additional python packages in the virtualenv, depending on the number of integrations and your NAS hardware, this could take awhile.

There's a way, mentioned in [this discussion comment](https://github.com/tprelog/iocage-homeassistant/discussions/54#discussioncomment-2129838) to gain some feedback about what's happening by setting the Home Assistant logger integration to `info` and having the log file print out in the terminal as things happen. The discussion itself does not relate to your issue but there is a boring, no audio, realtime video example showing how it works while reinstalling Home Assistant.

**Comment by aliekiddo** [comment]
> That's a really big jump for Home Assistant and this plugin. I imagine there are a few breaking changes you'll need to address with Home Assistant, including the required (yearly) python upgrade.

I didn't realise until recently that updating the plugin wasn't updating the Home Assistant version. I thought I had ran an upgrade in the Home Assistant console last week before this, but I know from another github thread I was running 2021.3.4 8 days ago - I just thought the TrueNAS version ran behind the current version.
I checked the solutions on this thread, the line to ignore OS version was present.

`HA_ARGS="--ignore-os-check --config ${homeassistant_config_dir}"`

>If you are using an external dataset for your configuration, I would suggest just doing a clean install of the plugin and remount your configuration there.

I am not using an external dataset unfortunately for a quick fix.

I started a plugin update at 08:50, I was getting a lot of 'lua script failed'

After a while TrueNAS kicked me out of the WebUI and I lost the ability to see what stage it was at. As far as I can see it's still sitting at 0% at 11:30, is it normal to take *this* long?

![2022-03-31 09 00 11](https://user-images.githubusercontent.com/7974754/161035641-ca2d6a57-5b75-4fbd-92dd-c9fe4d67e7de.jpg)

![Screenshot_20220331-103806](https://user-images.githubusercontent.com/7974754/161035706-39761a98-90b4-4b01-8af9-49fbca33213c.jpg)

**Comment by tprelog** [comment]
Ugh, sorry for the trouble @aliekiddo First and most important, **_Do not delete this jail_**. In any case we can still recover your configuration. While it is possible to further troubleshoot and repair this installation, I'm afraid it could be rather time intensive since the process can be some trial and error.

Can I instead offer to help you get setup with a new installation?

I can help you either move configuration to an external dataset now or transfer it directly to a new installation.

I was reading through the Home Assistant beta release notes for version `2022.4` and they are [adding a backup function to Home Assistant CORE](https://rc.home-assistant.io/blog/2022/03/30/release-20224/#backups-for-home-assistant-core--container-installs) - Not much details yet but I'm hoping there will be a web based restore method provided. This could potentially eliminate the need for using an external dataset.

**Comment by aliekiddo** [comment]
Thank you @tprelog and yes please. 
I figured a reinstall would be on the cards but spent a bit of time trying to see if there was anything else I could do.
I tried updating Python and reinstalling Home Assistant with the lines above but It kept saying it didn't recognise the command 'homeassistant' after the failed plugin update.
I tried to restore from several different snapshots but it that didn't seem to do anything, after it was done it didn't roll back to 2021.3.4.
I tried following this [installation guide](https://community.home-assistant.io/t/home-assistant-core-truenas-core-community-plugin/170542) to reinstall but I kept coming up against errors.
`ERROR: Failed building wheel for cryptography`

I've tried a couple of times to install a new HA plugin but it's got stuck at 50% and said the jail was corrupt.
A search on this error came up with running
`iocage clean -a`
but I'm not 100% sure if this is safe to do
